### PR TITLE
feat(Interoperability): Interoperability utilities for PyTorch/TensorFlow/Jax

### LINKED ISSUE #211: Add interoperability utils
There are a lot of operations we do in algorithms, like sum, that can be done using python `sum`, or specialized methods from `numpy` or `pytorch`. The specialized methods are generally more efficient, and so it makes a lot of sense to use them. However, we cannot ask the user to implement in each algorithm a set of `if ... elif ... else` to choose which implementation of `sum` to use based on the dtype.

An elegant solution would be to provide interoperability utilities (in a submodule `utils`, see also the reorganization of #153). These utilities could include common operations we encounter in algorithms like: sum, mean, copy, stack, zeros_like. The usage could then be:
```
from decent-bench.utils import interoperability as ip
ip.sum(elements)
```
where in the backend we check what dtype the elements have, and handle it specifically, e.g.:
```
def sum(elems):
    if numpy:
        return np.sum(np.stack(elems, axis=0), axis=0)
    elif pytorch:
        return torch.sum(torch.stack(elems, dim=0), dim=0)
    elif ...
        ...
    else:  # defaulting to python built-ins
        return sum(elems)
```

Copy method in particular could be implemented using `copy` package, since both `numpy` and `pytorch` expose `__copy__` and `__deepcopy__`.

Another operation that might require interoperability is indexing with a list, that is, passing a list of indices to a array object (numpy ndarray or pytorch tensor or ...). Indexing by list works for most packages except for python lists of lists, and we could have an interoperability tool that allows to always call `array[list of indices]`.

Of course this would be an optional utility, and users are free to implement their own code to only be comparible with e.g. numpy. But we will definitely be using this functionality.

### PR DESCRIPTION
**Pull request by Simpag**
Implements common operations using each frameworks native methods. To be discussed is behavior when a framework that is not supperted is used. Currently it tries to convert it to convert it into a numpy array and then applies the operator. This works fine in many cases since frameworks like scikit-learn, xgboost and lightgbm use numpy. Question is if we instead should raise an exception.

Changes to decentralized algorithms have been made so it uses the interoperability layer for most operations. Only operation left is initialization but that will be fixed in another issue.

The implementation requires no additional dependencies with negligible overhead over native framework usage.

closes #211

### CODE REVIEW

File: decent_bench/utils/interoperability.py [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+
+        Args:
+            array (ArrayLike): Input array (NumPy, torch, tf, jax) or nested container (list, tuple).
+
+        Returns:
+            ArrayLike: Array of random values in the same framework type as the input.
+
+        """
+        if isinstance(array, np.ndarray):
+            return cast("_T", np.random.default_rng().random(array.shape))
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/utils/interoperability.py]: Not wanted since rand_like can be called on an int tensor which is only used as a reference shape and framework type.

File: decent_bench/utils/interoperability.py [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+        if torch and isinstance(array, torch.Tensor):
+            return cast("_T", torch.rand_like(array))
+        if tf and isinstance(array, tf.Tensor):
+            shape = tf.shape(array)
+            return cast("_T", tf.random.uniform(shape))
+        if jnp and jax_random and isinstance(array, jnp.ndarray):
+            return cast(
+                "_T",
+                jax_random.uniform(jax_random.key(random.randint(0, 2**32 - 1)), shape=array.shape),
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/utils/interoperability.py]: Same as for numpy

File: decent_bench/utils/interoperability.py [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+
+        Args:
+            array (ArrayLike): Input array (NumPy, torch, tf, jax) or nested container (list, tuple).
+
+        Returns:
+            ArrayLike: Identity matrix in the same framework type as the input.
+
+        """
+        if isinstance(array, np.ndarray):
+            return cast("_T", np.eye(*array.shape[-2:], dtype=array.dtype, device=array.device))
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/utils/interoperability.py]: Added in numpy 2.0.0

File: decent_bench/utils/interoperability.py [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+
+        if isinstance(like, np.ndarray):
+            return cast("_T", array)
+        if torch and isinstance(like, torch.Tensor):
+            return cast("_T", torch.from_numpy(array).to(device))
+        if tf and isinstance(like, tf.Tensor):
+            with tf.device(device):
+                return cast("_T", tf.convert_to_tensor(array))
+        if jnp and isinstance(like, jnp.ndarray):
+            return cast("_T", jnp.array(array, device=device))
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/utils/interoperability.py]: Same as for jnp.eye

File: pyproject.toml [hunk 5]
```diff
@@ -34,8 +34,14 @@ dev = [
   "ruff",
   "scipy-stubs",
   "sphinx",
+  "sphinx-toolbox",
   "types-networkx",
   "types-tabulate",
+  "torch",
+  "torchvision",
+  "tensorflow[and-cuda]",
```
[the remark this answers is not in the document]
Comment (Simpag) [reply pyproject.toml]: Its the way they say to [install](https://www.tensorflow.org/install) it.

File: decent_bench/utils/interoperability.py [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+            return cast(
+                "_T",
+                jax_random.uniform(jax_random.key(random.randint(0, 2**32 - 1)), shape=array.shape),
+            )
+        if isinstance(array, (list, tuple)):
+            np_array = Interoperability.to_numpy(array)
+            random_array = np.random.default_rng().random(np_array.shape)
+            return cast("_T", type(array)(random_array.tolist()))
+        np_array = Interoperability.to_numpy(array)
+        return cast("_T", np.random.default_rng().random(np_array.shape))
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/utils/interoperability.py]: Same as for numpy case.

File: decent_bench/utils/interoperability.py [hunk 7]
```diff
@@ -0,0 +1,591 @@
...
+    if tf and isinstance(array, tf.Tensor):
+        return cast("T", tf.identity(array))
+    if jnp and isinstance(array, jnp.ndarray):
+        return cast("T", jnp.array(array, copy=True))
+    return deepcopy(array)
+
+
+def stack[T: ArrayLike](arrays: Sequence[T], dim: int = 0) -> T:
+    """
+    Stack a sequence of arrays along a new axis.
```
Comment (nicola-bastianello) [inline decent_bench/utils/interoperability.py]: "along the specified axis" rather than "along a new axis"? because it might be an existing axis right? e.g. if I stack column vectors in a column

File: decent_bench/utils/interoperability.py [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+    if isinstance(array, (list, tuple)):
+        np_array = to_numpy(array)
+        return cast("T", type(array)(np.ones_like(np_array).tolist()))
+
+    raise TypeError(f"Unsupported framework type: {type(array)}")
+
+
+def rand_like[T: ArrayLike](array: T) -> T:
+    """
+    Create an array of random values with the same shape as the input.
```
Comment (nicola-bastianello) [inline decent_bench/utils/interoperability.py]: should specify that the random values are uniformly drawn from [0,1]

Comment (Simpag) [reply decent_bench/utils/interoperability.py]: Its easy to add a normal one

File: decent_bench/utils/interoperability.py [hunk 9]
```diff
@@ -503,21 +505,73 @@ def rand_like[T: ArrayLike](array: T) -> T:
...
+        random_array = np.random.default_rng().random(
+            size=np_array.shape,
+            dtype=np_array.dtype,
+        )
+        return cast("T", type(array)(random_array.tolist()))
+
+    raise TypeError(f"Unsupported framework type: {type(array)}")
+
+
+def randn_like[T: ArrayLike](array: T) -> T:
```
Comment (nicola-bastianello) [inline decent_bench/utils/interoperability.py]: would it be possible to have the kwargs "loc" and "scale" so that users can customize mean and variance of the normal distribution? this is going to be very useful if we want to calibrate e.g. the intensity of the communication noise

Comment (Simpag) [reply decent_bench/utils/interoperability.py]: PyTorch and JAX don't have parameters for mean and var. We could add mean and std and apply mean + random_norm * std
